specific variant targets for NAZE rev5, rev6

christianrauch opened the pull request:
This commit provides variant targets for rev5 and rev6 of NAZE to use the available flash more efficiently. If no specific revision is given, the generic NAZE target will be built.
Only `NAZE_rev6` has been tested on real hardware for now.
I am rebasing this this commit on tag `CLFL_v2.0.2` since NAZE on `master` (069e837f106af1d13700c0fcc8b4ef723bbfd1e1) is not booting.

mikeller commented on src/main/target/NAZE/NAZE_rev5.mk:
```diff
@@ -0,0 +1 @@
+# NAZE rev5, STM32F103CB, MPU6050, HMC5883L, MS5611
```
Target names are all caps.

mikeller commented on src/main/target/NAZE/target.h:
```diff
@@ -74,36 +82,66 @@
+#endif
+
 
+#ifdef NAZE_GENERIC
```
These conditionals should not be in here. Defines that are not used are simply ignored, making them conditional just clutters up the code. If anything, move the `GYRO_X_ALIGN` defines up next to the `USE_GYRO_X` defines.

mikeller commented on src/main/target/NAZE/target.h:
```diff
@@ -74,36 +82,66 @@
 #define USE_ACC_BMA280
 #define USE_ACC_MMA8452
+#endif
+#if defined(NAZE_GENERIC) || defined(NAZE_rev5)
```
Why are you adding this to `NAZE_GENERIC`? The 3050's datasheet says that it does not have an accelerometer: https://www.invensense.com/products/motion-tracking/3-axis/mpu-3050/

christianrauch replied:
I am not sure what you mean here. Only the Gyro of the MPU 3050 is used (as in the original `target.h`). This check activates the `USE_ACC_MPU6050` for rev5 targets.

mikeller replied:
My bad. I had not realised at that point that your aim is to make NAZE_GENERIC support all boards, including revs 5 and 6.

mikeller commented:
Thanks for making the improvements. I would still like an explanation as to what benefits this change will bring. As it is now, I don't see what benefits would justify the additional effort required to maintain two more targets.

christianrauch commented:
I addressed some of the comments.
My statement:

> use the available flash more efficiently

relates to the fact that on rev5 and rev6 targets we do not need to include all sensors, but only those that are actually on the hardware. The remaining space can instead be used for additional features, e.g. the magnetometer and the `USE_RX_MSP`.

I assume that people that are using cleanflight are aware of which version of the NAZE hardware they are using. If not, they can still use the generic `NAZE` target as before, which will work on all NAZE revisions. Users that want to use more features can refer to the specific firmware version.
Determining your NAZE revision is still easier than checking out the sourcecode and manually patching the `target.h`.

mikeller commented:
> relates to the fact that on rev5 and rev6 targets we do not need to include all sensors, but only those that are actually on the hardware. The remaining space can instead be used for additional features, e.g. the magnetometer and the USE_RX_MSP.

Magnetometer is already active in the current configuration. RX_MSP is a feature that is 'for testing only', so it shouldn't be used by users not actually working on the firmware. Both of them are not really justifications for adding more targets.

You have to understand that support for NAZE and the other F1 targets is kept in cleanflight / Betaflight just so that owners of craft with these boards can 'dip their toe' into recent versions of the firmware, before they have to commit to buying new hardware and modifying their craft. The real benefits of recent versions of the firmware (faster loop times, DShot, more features,...) will only be available to them once they upgrade. There are regular calls from developers for F1 support to be dropped altogether, since it creates extra work, so justifying adding even more work to this is not really possible.

christianrauch commented:
At least on the NAZE [target.h](https://github.com/cleanflight/cleanflight/blob/master/src/main/target/NAZE/target.h#L104), the magnetometer is commented out, so it is not used at all.

Why is `USE_RX_MSP` only for people working on the firmware? As I understand, it allows you to control the FC via `MSP_SET_RAW_RC` messages, e.g. through a companion PC. I see this as a user feature and not a developer feature. And even if `USE_RX_MSP` is defined, it is not active by default.

I am actually surprised that dropping F1 support is actively discussed. I thought that F1 flight controller are widely used, since they are cheap and easy to use.

I still think that using the magnetometer and controlling the flight controller via `MSP_SET_RAW_RC` are useful features. But if there are still strong rejections against providing these specific targets with additional features (compared to the generic version), feel free to close this PR. I will then recommend clean version 1.14.2 to people that want to use the magnetometer and control the flight controller via `MSP_SET_RAW_RC` messages.

mikeller commented:
> At least on the NAZE target.h, the magnetometer is commented out, so it is not used at all.

For some reason that is not what you are basing your PR on.

> Why is USE_RX_MSP only for people working on the firmware? 

I guess part of it is that it is not as extensively developed and tested for reliability as the other RX types, since it is not intended to be used with a craft with props on. Users wanting to do so can - but it's at their own risk of getting attacked by a blender gone mad.

I am actually surprised that dropping F1 support is actively discussed. I thought that F1 flight controller are widely used, since they are cheap and easy to use.

Why are you surprised? The technology has moved on, and as much as it would be nice if today's float based PID controller could run at a reasonable speed on hardware without float support, that's just not the reality. To exacerbate this, the amount of useful features that has been added, plus the improvements that have been made to existing features simply have no way to fit into the 128K flash that F1 has.

> I will then recommend clean version 1.14.2 to people that want to use the magnetometer and control the flight controller via MSP_SET_RAW_RC messages.

That is ok. This is open source, and people are free to use whatever version they like. If they are happy with the performance and features they get out of 1.14.2, and that there is little support and no bug fixes available, then there is not much reason for them to upgrade. As I said before, you have to understand that F1 support on 2.0 and on is only kept in as a kind of 'technology preview', and is not intended to be made into more than this.

@hydra: Do you agree to closing this?

martinbudden commented:
@christianrauch , I don't like to discourage people who make contributions to the codebase, however as is pointed out earlier in the thread, there is a maintenance cost to these changes. Whether it is worth bearing the maintenance cost depends on the utility of the features added. I would point out:

1. I haven't seen people asking for RX_MSP or magnetometer support on NAZE
2. On board magnetometer support is actually virtually useless anyway, since the high currents near the flight controller render its output unreliable (see various discussions around iNav)
3. If the user wants magnetometer supported flight modes, then iNav is a better choice than betaflight.

>  I will then recommend clean version 1.14.2 to people that want to use the magnetometer
As stated about, it is better to recommend iNav to people who want to use the magnetometer.

Having said all the above, please don't be discouraged from making future contributions.

christianrauch commented:
The primary feature I would like to have enabled at compile time is actually `USE_RX_MSP`.
I am developing a C++ library for communicating with flight controllers over MSP. I would say that most people that are using this library are interested in controlling a FC with an onboard PC using the `MSP_SET_RAW_RC` message. If this feature is removed at compile time, people need to manually check out the source code and add/remove a couple of `#defines` to have this feature enabled again.

martinbudden commented:
I'm much more sympathetic to enabling `USE_RX_MSP`, especially if there is (or becomes) available a C++ library that enables controlling flight controllers over MSP.

Turing on `USE_RX_MSP` cost 224 bytes on NAZE and currently fits in the build, so this is a currently a viable option.

mikeller commented:
Totally agreed with @martinbudden on this. Just be aware that it will be a prime candidate to be eliminated if space needs to be freed up to make the firmware fit onto F1 targets in the future.

Also, may I suggest that you get yourself an F3 based board for testing? They are available from USD 10 from China, and the amount of flash space is much less of a problem with them.

christianrauch commented:
There is a new PR https://github.com/cleanflight/cleanflight/pull/2819 to just activate `USE_RX_MSP`.

I am developing a C++ library for MSP communication at: https://github.com/christianrauch/msp. It is supposed to support all firmwares that use MSP communication (MultiWii, Cleanflight, Betaflight).
